M2/Macaulay2/packages/Msolve.m2

```diff
@@ -0,0 +1,527 @@
+		  Email => "leykin@math.gatech.edu", 
+		  HomePage => "https://antonleykin.math.gatech.edu/"}},
+    	Headline => "An interface to the msolve package (https://msolve.lip6.fr/) which computes Groebner Basis and does real root isolation",
+    	AuxiliaryFiles => true,
+	DebuggingMode => false,       
+	OptionalComponentsPresent => msolvePresent
+	);
+msolveEXEwithOpts:=(opt)->"msolve -v "|opt#"level of verbosity"|" -t "|toString(opt#"number of threads");
+export{
+    "msolveGB",
+    "msolveSaturate",
+    "msolveEliminate",
+    "msolveRUR",
+    "msolveLeadMonomials",
+    "msolveRealSolutions"
+    }
+importFrom_Core { "raw", "rawMatrixReadMsolveFile" }
+
+use'readMsolveOutputFile := true;
+readMsolveOutputFile = method()
+readMsolveOutputFile(Ring,String) := Matrix => (R,mOut) -> if use'readMsolveOutputFile 
+    then map(R, rawMatrixReadMsolveFile(raw R, mOut)) else (
+	msolveOut := get mOut;
+    	moutStrL:=separate(",",first separate("[]]",last separate("[[]",msolveOut)));
+    	--the line below should be replaced by a call to the C-function to parse the string
+    	M2Out:=for s in moutStrL list value(s);
+    	matrix {M2Out};
+    	)
+
+inputOkay=method(TypicalValue=>Boolean);
+inputOkay(Ideal):=I->(
+    R:=ring I;
+    if not instance(R,PolynomialRing) then (print "input must be in a polynomial ring over a field";return false;);
+    kk:=coefficientRing(R);
+    if instance(kk,InexactFieldFamily) then(
+        print "input must be over the rationals or a (prime) finite field of chacteristic between 2^16 and 2^32";
+	return false;
+	);
+    if char(kk)>0 then (
+	if (char(kk)<2^16) or (char(kk)>2^32) then(
+	    print "input must be over the rationals or a (prime) finite field of chacteristic between 2^16 and 2^32";
+	    return false;
+	    );
+	);
+    return true;
+    );
+msolveGB=method(TypicalValue=>Matrix, Options=>{"level of verbosity"=>0, "number of threads"=>maxAllowableThreads});
+msolveGB(Ideal):=opt->I->(
+    if not inputOkay(I) then return 0;
+    mIn:=temporaryFileName()|".ms";
+    R:=ring I;
+    kk:=coefficientRing(R);
+    gR:=toString(gens R);
+    l1:=substring(1,#gR-2,gR);
+    l2:=char R;
+    gI:=toString flatten entries gens I;
+    if (isField(kk) and (char(kk)==0)) then gI=replace("[)(]","",gI);
+    Igens:=replace(" ",""|newline,substring(1,#gI-2,gI));
+    inStr:=l1|newline|l2|newline|Igens;
+    mIn<<inStr<<close;
+    mOut:=temporaryFileName()|".ms";
+    << "msolve input file is called: " << mIn << endl;
+    << "msolve output file is called: " << mOut << endl;
+    callStr:=msolveEXEwithOpts(opt)|" -g 2 -f "|mIn|" -o "|mOut;
+    run(callStr);
+    <<"done msolve"<<endl;
+    msolGB:=readMsolveOutputFile(R, mOut);
+    return gens forceGB msolGB;
+    );
+msolveLeadMonomials=method(TypicalValue=>Matrix, Options=>{"level of verbosity"=>0, "number of threads"=>maxAllowableThreads});
+msolveLeadMonomials(Ideal):=opt->(I)->(
+    if not inputOkay(I) then return 0;
```

Comment:
Same, throwing an error. We can add hooks for `leadMonomial` and add this as a hook, which can be specified using the `Strategy`.

Reply:
msolve always returns monic monomials with the GB option which computes the lead term. So this seems like a reasonable name?

Reply:
Aren't the lead terms of a Grobner basis over a field are always monic? e.g. in M2 `leadTerm gb I` always returns monic monomials also, so I think it's important to align the methods with M2's native ones.

Reply:
leadTerm groebnerBasis I
and
 leadTerm I
both return a matrix with the coefficients included (and are not necessarily monic). 

And for me leadTerm gb I does not nessarily return monic monomials (of course you can just throw away the coefficients, but that is not what it does, and leadTerm seems more correct when you keep the coefficients... the ideal is of course the same either way). 

Anyway, to me the msolve GB is closer aligned to the "groebnerBasis" command rather than the "gb" command, which was why I opted to do it this way, i.e. if it was msolveLeadTerm I would want its output to match  leadTerm groebnerBasis I, but it won't. 

Also msolve only has grevlex and block grevlex (as an elimination order) defined, so in some sense I am not sure we  want to align it too closely to the M2 commands since it has more limited capabilities in some sense.

Reply:
Ah interesting, it seems like this is actually field dependent:
```m2
i21 : R = ZZ[x,y]; leadTerm gb ideal (2*x+3*y)

o22 = | 2x |

              1      1
o22 : Matrix R  <-- R

i23 : R = ZZ/5[x,y]; leadTerm gb ideal (2*x+3*y)

o24 = | x |

              1      1
o24 : Matrix R  <-- R
```
@mikestillman is this a bug?